### PR TITLE
Fix ASan report about alloc/dealloc mismatch when compiled with C++23

### PR DESCRIPTION
**Pull request by waahm7**
*Issue #, if available:*

*Description of changes:*
Duplicate of #629 to fix lint and submodules

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: include/aws/crt/StlAllocator.h
Comment (graebm) [inline include/aws/crt/StlAllocator.h:1]: If `std::allocator<T>` doesn't have any virtual functions ... does that mean nothing should inherit from it?

Comment (waahm7) [reply include/aws/crt/StlAllocator.h:1]: Discussed offline, we will backlog this for now.

File: include/aws/crt/StlAllocator.h [hunk 1]
```diff
@@ -51,6 +51,10 @@ namespace Aws
                 return static_cast<RawPointer>(aws_mem_acquire(m_allocator, n * sizeof(T)));
             }
 
+#if _LIBCPP_STD_VER > 20
```
Comment (DmitriyMusatkin) [inline include/aws/crt/StlAllocator.h:54]: do we really need #if here? 
does it really matter if we expose it in all versions of cpp standard?

Comment (waahm7) [reply include/aws/crt/StlAllocator.h]: std::allocation_result was added in _LIBCPP_STD_VER > 20

**Review by DmitriyMusatkin — APPROVED** [review]

**Pull request merged**